### PR TITLE
Fixes in internal logic of sending packets

### PR DESCRIPTION
**Pull request by aggarw13**
**This PR fixes 2 different issues, one for `sendPacket` and one for `sendPublish` internal functions,** both related to sending packets over the network.

#### Issue 1
For `sendPacket`, there is a possibility of encountering an infinite loop if the `Transport_Send_t` always returns zero bytes sent. The PR addresses this issue by adding a timeout logic on retrying zero send attempts. The timeout can be configured through the `MQTT_SEND_RETRY_TIMEOUT_MS` macro. 

#### Issue 2
For the `sendPublish` function (that calls the `sendPacket` function to send the complete packet over the network), there is a bug of treating partial or zero bytes return value of `sendPacket` as success. This PR updates the logic to treat partial or zero transmission of packets as failure in the call to `MQTT_Publish` API.

### CODE REVIEW

File: test/unit-test/core_mqtt_utest.c [hunk 1]
```diff
@@ -1949,17 +2011,21 @@ void test_MQTT_ProcessLoop_Timer_Overflow( void )
...
     setupTransportInterface( &transport );
+    setupNetworkBuffer( &networkBuffer );
 
     networkBuffer.size = 1000;
     incomingPacket.type = MQTT_PACKET_TYPE_PUBLISH;
     incomingPacket.remainingLength = MQTT_SAMPLE_REMAINING_LENGTH;
 
     globalEntryTime = UINT32_MAX - MQTT_OVERFLOW_OFFSET;
-    expectedFinalTime = MQTT_TIMER_CALLS_PER_ITERATION * numIterations - MQTT_OVERFLOW_OFFSET;
+    expectedFinalTime = globalEntryTime + ( numIterations * MQTT_TIMER_CALLS_PER_ITERATION ) + 1;
```
Comment (muneebahmed10) [inline test/unit-test/core_mqtt_utest.c:2028]: Nit: Why this change? I liked the way it looked before, where we calculate the expected time independently. If we're testing if an overflow calculation works as expected, it looks a little weird to me that we're doing the exact same overflow calculation here (I know since they're unsigned it's not really an "overflow" and addition is well-defined, and that these two expressions will always be identical for uint32_t, but you know what I mean).

File: test/unit-test/core_mqtt_utest.c [hunk 2]
```diff
@@ -1949,17 +2011,21 @@ void test_MQTT_ProcessLoop_Timer_Overflow( void )
     MQTTPublishState_t publishState = MQTTPubAckSend;
     MQTTPublishState_t ackState = MQTTPublishDone;
     uint8_t i = 0;
-    uint8_t numIterations = ( MQTT_TIMER_OVERFLOW_TIMEOUT_MS / MQTT_TIMER_CALLS_PER_ITERATION ) + 1;
+    uint8_t numIterations = ( ( MQTT_TIMER_OVERFLOW_TIMEOUT_MS % MQTT_TIMER_CALLS_PER_ITERATION ) == 0 ) ?
+                            ( MQTT_TIMER_OVERFLOW_TIMEOUT_MS / MQTT_TIMER_CALLS_PER_ITERATION ) :
+                            ( ( MQTT_TIMER_OVERFLOW_TIMEOUT_MS / MQTT_TIMER_CALLS_PER_ITERATION ) + 1 );
```
Comment (muneebahmed10) [inline test/unit-test/core_mqtt_utest.c:2016]: Is this just ceil( a / b )?  I believe something like `( MQTT_TIMER_OVERFLOW_TIMEOUT_MS + MQTT_TIMER_CALLS_PER_ITERATION - 1 ) / MQTT_TIMER_CALLS_PER_ITERATION` would be equivalent and somewhat shorter, right? And a comment explaining that this is just a / b rounded up?

File: source/core_mqtt.c [hunk 3]
```diff
@@ -1386,7 +1408,7 @@ static MQTTStatus_t sendPublish( MQTTContext_t * pContext,
                             pContext->networkBuffer.pBuffer,
                             headerSize );
 
-    if( bytesSent < 0 )
+    if( bytesSent < ( int32_t ) headerSize )
```
Comment (muneebahmed10) [inline source/core_mqtt.c:1411]: Are you going to make the same change in the other places where `sendPacket` is called, or will that be a separate PR?

File: source/include/core_mqtt_config_defaults.h [hunk 4]
```diff
@@ -129,6 +129,28 @@
...
+ *
+ * When sending an MQTT packet, the transport send function may be called multiple
+ * times until all of the required number of bytes are sent.
+ * This timeout represents the maximum duration that is allowed for no data
+ * transmission over the network through the transport send function.
+ *
+ * If the timeout expires, the #MQTT_ProcessLoop and #MQTT_ReceiveLoop functions
+ * return #MQTTSendFailed.
+ *
+ * <b>Possible values:</b> Any positive integer up to SIZE_MAX. Recommended to
```
Comment (muneebahmed10) [inline source/include/core_mqtt_config_defaults.h:145]: Now that I think about it, why do we say up to SIZE_MAX if it will only be compared to a uint32_t? On our dev systems, isn't that 2^ 64 - 1? Shouldn't this (and others) be "any positive 32 bit integer"?

File: source/interface/transport_interface.h [hunk 5]
```diff
@@ -183,7 +183,10 @@ typedef int32_t ( * TransportRecv_t )( NetworkContext_t * pNetworkContext,
  * @param[in] pBuffer Buffer containing the bytes to send over the network stack.
  * @param[in] bytesToSend Number of bytes to send over the network.
  *
- * @return The number of bytes sent or a negative error code.
+ * @return The number of bytes sent or a negative value to indicate error.
+ * If no data is transmitted over the network due to a full TX buffer and
```
Comment (muneebahmed10) [inline source/interface/transport_interface.h:187]: Nit: We used `@note`here when making a similar comment for TransportRecv_t. Also I think there would have to be an empty line between the `@return` and `@note` blocks, judging from the other doxygen comments

File: test/unit-test/core_mqtt_utest.c (hunk 1 above)
Comment (aggarw13) [reply test/unit-test/core_mqtt_utest.c:2028]: This is correcting the calculation of the expected time counter after the call to `MQTT_ProcessLoop`. The `+1` is for the call made to `getTime` before all iterations of the loop in the API.

File: source/core_mqtt.c (hunk 3 above)
Comment (aggarw13) [reply source/core_mqtt.c:1411]: Separate PR to avoid scope screep in this PR.

File: source/include/core_mqtt_config_defaults.h (hunk 4 above)
Comment (aggarw13) [reply source/include/core_mqtt_config_defaults.h:145]: Sure, good point. Made the change

File: source/interface/transport_interface.h (hunk 5 above)
Comment (aggarw13) [reply source/interface/transport_interface.h]: Made the API doc for send and receive consistent.

File: test/unit-test/core_mqtt_utest.c (hunk 2 above)
Comment (aggarw13) [reply test/unit-test/core_mqtt_utest.c]: Good suggestion, have made the suggested changes.

File: test/unit-test/core_mqtt_utest.c (hunk 1 above)
Comment (muneebahmed10) [reply test/unit-test/core_mqtt_utest.c:2034]: What's the correction? This value is the same as before isn't it? 
`(UINT32_MAX - MQTT_OVERFLOW_OFFSET) + (numIterations * MQTT_TIMER_CALLS_PER_ITERATION ) + 1` is the same as 
`((0) - MQTT_OVERFLOW_OFFSET) + (numIterations * MQTT_TIMER_CALLS_PER_ITERATION)` or
`MQTT_TIMER_CALLS_PER_ITERATION * numIterations - MQTT_OVERFLOW_OFFSET`

Also, basing `expectedFinalTime` off of `globalEntryTime` results in a calculation that always yields the final value, regardless if the counter rolls over back to 0 or not, which is what this test seems to be testing (e.g. If someone accidentally changed the value of `globalEntryTime` so there's no longer a rollover, this test would still pass).

Ultimately it doesn't matter, since the value is the same as it was before, but that's why this change looked weird to me

File: source/core_mqtt.c [hunk 6]
```diff
@@ -628,20 +640,30 @@ static int32_t sendPacket( MQTTContext_t * pContext,
...
+        }
+        else
+        {
+            /* No bytes were sent over the network. */
+            timeSinceLastSendMs = calculateElapsedTime( pContext->getTime(), lastSendTimeMs );
+
+            /* Check for timeout if we have been waiting to send any data over the network. */
+            if( timeSinceLastSendMs >= MQTT_SEND_RETRY_TIMEOUT_MS )
+            {
+                LogError( ( "Unable to send packet: Timed out in sending any data through transport send." ) );
```
Comment (muneebahmed10) [inline source/core_mqtt.c:655]: It sounds like it uses "send" one too many times
```suggestion
                LogError( ( "Unable to send packet: Timed out in transport send." ) );
```

File: source/core_mqtt.c [hunk 7]
```diff
@@ -755,7 +777,7 @@ static int32_t recvExact( const MQTTContext_t * pContext,
             /* Check for timeout if we have been waiting to receive any byte on the network. */
             if( timeSinceLastRecvMs >= MQTT_RECV_POLLING_TIMEOUT_MS )
             {
-                LogError( ( "Time expired while receiving packet." ) );
+                LogError( ( "Unable to receive packet: Timed out in receiving any data from transport recv." ) );
```
Comment (muneebahmed10) [inline source/core_mqtt.c:780]: Same
```suggestion
                LogError( ( "Unable to receive packet: Timed out in transport recv." ) );
```

**Review by muneebahmed10 — APPROVED** [review]

**Review by muneebahmed10 — APPROVED** [review]

File: test/unit-test/core_mqtt_utest.c (hunk 1 above)
Comment (aggarw13) [reply test/unit-test/core_mqtt_utest.c:2034]: >Also, basing expectedFinalTime off of globalEntryTime results in a calculation that always yields the final value, regardless if the counter rolls over back to 0 or not,

This is the main motivation from a resiliency and readability perspective.

**Review by gkwicker — APPROVED** [review]

File: test/unit-test/core_mqtt_utest.c (hunk 1 above)
Comment (muneebahmed10) [reply test/unit-test/core_mqtt_utest.c:2034]: I was saying it's less resilient since before, accidentally setting `globalEntryTime` incorrectly could be caught due to the test failing. Now, due to how `expectedFinalTime` is calculated from `globalEntryTime`, the test will always pass regardless of the initial value of the entry time
